firedrake/halo.py

```diff
@@ -103,17 +103,7 @@ def __init__(self, dm, section, comm):
 
     @utils.cached_property
     def sf(self):
-        lsec = self.dm.getDefaultSection()
-        gsec = self.dm.getDefaultGlobalSection()
-        self.dm.createDefaultSF(lsec, gsec)
```

Comment:
Could you instead create a global section that includes the constraint dofs via `PetscSectionCreateGlobalSection` and then continue with `dm.createDefaultSF(lsec, gsec)`? (Although I see the petsc4py wrapper needs to expose the `includeConstraints` argument first)

Reply:
The global section not including the constraint DoFs is actually the one which represents the layout of the reduced (shorter) global vector (that we use in solve), so I think it is the one our section dm should carry. Or did you mean we could make a new global section that includes constraint DoFs when we make halo exchange sf?

By the way, the idea in this PR https://github.com/firedrakeproject/firedrake/pull/3426 is that we compute `renumbering` so that the owned constrained DoFs will be pushed to the end of the "core + owned" block so that we can make reduced global vector by concatenating the "core + owned - constrained" block. We do not care where constrained DoFs appear in the halo.

Reply:
Yes, that is the idea.